crates/scarb-api/src/lib.rs

```diff
@@ -75,64 +78,121 @@ fn artifacts_for_package(path: &Utf8Path) -> Result<StarknetArtifacts> {
-            .find_map(|test_type| check_path(&starknet_artifacts_file_name(Some(test_type))))
-    } else {
-        check_path(&starknet_artifacts_file_name(None))
-    }
+) -> Option<ContractArtifactData> {
+    let path = format!("{target_name}.starknet_artifacts.json");
+    let path = build_target_path(target_dir, &path);
+    path.map(|path| ContractArtifactData {
+        path,
+        test_type: None,
+    })
 }
 
 /// Get the map with `StarknetContractArtifacts` for the given package
 pub fn get_contracts_artifacts_and_source_sierra_paths(
     metadata: &Metadata,
-    package: &PackageId,
-    profile: Option<&str>,
+    target_dir: &Utf8Path,
+    package: &PackageMetadata,
     use_test_target_contracts: bool,
 ) -> Result<HashMap<String, (StarknetContractArtifacts, Utf8PathBuf)>> {
-    let target_name = target_name_for_package(metadata, package)?;
-    let target_dir = target_dir_for_workspace(metadata);
-    let maybe_contracts_path = get_starknet_artifacts_path(
-        &target_dir,
-        &target_name,
-        profile.unwrap_or(metadata.current_profile.as_str()),
-        use_test_target_contracts,
-    );
-
-    let map = match maybe_contracts_path {
-        Some(contracts_path) => load_contracts_artifacts_and_source_sierra_paths(&contracts_path)?,
-        None => HashMap::default(),
+    let contracts_paths = if use_test_target_contracts {
+        let test_targets = test_targets_by_name(package);
+        get_starknet_artifacts_paths_from_test_targets(target_dir, &test_targets)
+    } else {
+        let target_name = target_name_for_package(metadata, &package.id)?;
+        get_starknet_artifacts_path(target_dir, &target_name)
+            .into_iter()
+            .collect()
     };
 
-    Ok(map)
+    if contracts_paths.is_empty() {
+        Ok(HashMap::default())
+    } else {
+        load_contracts_artifacts(&contracts_paths)
+    }
+}
+
+fn load_contracts_artifacts(
+    contracts_paths: &[ContractArtifactData],
+) -> Result<HashMap<String, (StarknetContractArtifacts, Utf8PathBuf)>> {
+    if contracts_paths.is_empty() {
+        return Ok(HashMap::new());
+    }
+
+    // TODO use const
+    let base_artifacts = contracts_paths
+        .iter()
+        .find(|paths| paths.test_type == Some("integration".to_string()))
+        .unwrap_or(
+            contracts_paths
+                .first()
+                .expect("Must have at least one value because of the assert above"),
+        );
+
+    let other_artifacts: Vec<&ContractArtifactData> = contracts_paths
+        .iter()
+        .filter(|path| path != &base_artifacts)
+        .collect();
+
+    let mut base_artifacts =
+        load_contracts_artifacts_and_source_sierra_paths(&base_artifacts.path)?;
+
+    for artifact in other_artifacts {
+        let artifact = load_contracts_artifacts_and_source_sierra_paths(&artifact.path)?;
+        for (key, value) in artifact {
+            base_artifacts.entry(key).or_insert(value);
+        }
+    }
```

Comment:
Do I understand correctly, this is only meant to be a fallback? 
If so, isn't it extremely wasteful, that it compiles all contract files with USC before checking if you even need them?

Reply:
Btw. I still think it should be possible to compile all contract files in parallel with multiple usc processes.

Reply:
We can use `par_iter` to compile these in parallel, it should be safe.